examples/step-93/step-93.cc

```diff
@@ -0,0 +1,840 @@
+    // also put the constraints for boundary values on both the $u$
+    // and $\lambda$ variables.
+    //
+    // Because the nonlocal degrees of freedom use discontinuous
+    // elements, they do not contribute to boundary values
+    // (discontinuous elements do not have degrees of freedom
+    // logically located on the boundary that could be interpolated)
+    // and we do not need to exclude these solution components
+    // explicitly when calling
+    // VectorTools::interpolate_boundary_values()
+    constraints.clear();
+    DoFTools::make_hanging_node_constraints(dof_handler, constraints);
+    VectorTools::interpolate_boundary_values(dof_handler,
+                                             0,
+                                             Functions::ZeroFunction<dim>(3),
+                                             constraints);
+    constraints.close();
+
+    // The remainder of the function deals with building the sparsity
+    // pattern. It consists of two parts: The entries that result from
+    // the usual integration of products of shape functions, and then
+    // the entries that result from integrals that contain nonlocal
+    // degrees of freedom (which one can think of as associated with
+    // shape functions that are constant across the entire
+    // domain). The first part is easily built using standard tools:
+    DynamicSparsityPattern dsp(dof_handler.n_dofs(), dof_handler.n_dofs());
+    DoFTools::make_sparsity_pattern(dof_handler, dsp, constraints, false);
+
+    // The other part is more awkward. We have matrix entries that
+    // result from terms such as $\int_{\Omega}\varphi_j f_k$ where
+    // each $\varphi_j$ is a shape function associated to $\lambda$
+    // and each $f_k$ is a
+    // characteristic function of a part of the domain. As a
+    // consequence, we end up with a nonzero matrix entry $A_{jk}$
+    // (along with its transpose $A_{kj}$) if there is overlap between
+    // $\varphi_j$ and $f_k$. In practice, because we will use
+    // quadrature, this means that we end up with a quadrature point
+    // on a cell on which $\varphi_j$ lives and at which $f_k$ is not
+    // zero. (We will implicitly assume that a shape function that
+    // lives on the current cell is nonzero at all quadrature points
+    // -- an assumption that is generally true unless one chooses
+    // specific quadrature formulas.) Determining which sparsity
+    // pattern entries we need to add then essentially comes down to
+    // "simulating" what would happen if we actually computed
+    // integrals and which matrix entries would end up being non-zero
+    // in the process. As a consequence, the following code's general
+    // structure looks very similar to what we will do for the
+    // nonlocal contributions in the `assemble_system()` function
+    // below.
+    //
+    // To get this started, we create a dummy `hp_fe_values` that we
```

Comment:
```suggestion
    // To get this started, we create a `hp_fe_values` that we
```
It is a dummy object but sometimes that word trips students up (i.e., they have not heard the phrase "dummy variable" before) - lets just call it what it is.